src/rover/camera.cpp

```diff
@@ -0,0 +1,229 @@
+
+    // Query for capability.
+    if (ioctl(session->fd, VIDIOC_QUERYCAP, &cap) != 0) {
+        return Error::QUERY_CAPABILITIES;
+    }
+
+    // Keep the camera's name just in case we want this later.
+    strcpy(session->name, (char*) cap.card);
+
+    // Also keep the camera's hardware location.
+    strcpy(session->hardware_location, (char*) cap.bus_info);
+
+    // Does the camera support video capture?
+    if (!(cap.capabilities & V4L2_CAP_VIDEO_CAPTURE)) {
+        return Error::NO_VIDEOCAPTURE;
+    }
+
+    // Does this camera support video streaming?
+    if (!(cap.capabilities & V4L2_CAP_STREAMING)) {
+        return Error::NO_STREAMING;
+    }
+
+    // Allocate a format struct for the following calls.
+    struct v4l2_format fmt;
+    memset(&fmt, 0, sizeof(fmt));
+
+    // Instruct V4L2 to ask for video format.
+    fmt.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+
+    // Let V4L2 fill it in based on what is currently set.
+    // This is so that we only need to set relevant values.
+    if (ioctl(session->fd, VIDIOC_G_FMT, &fmt) != 0) {
+        return Error::QUERY_FORMAT;
+    }
+
+    // Then set the pixel format we want.
+    // Also set our desired resolution.
+    fmt.fmt.pix.pixelformat = PIXEL_FORMAT;
+    fmt.fmt.pix.width = session->width;
+    fmt.fmt.pix.height = session->height;
+
+    // Return the format with our changes.
+    if (ioctl(session->fd, VIDIOC_S_FMT, &fmt) != 0) {
+        return Error::SET_FORMAT;
+    }
+
+    // Check the returned pixel format.
+    // If the camera does not support the pixel format we desire,
+    // it will return whatever it can use.
+    if (fmt.fmt.pix.pixelformat != PIXEL_FORMAT) {
+        return Error::UNSUPPORTED_FORMAT;
+    }
+
+    // Check the returned dimensions.
+    if (fmt.fmt.pix.width != session->width || fmt.fmt.pix.height != session->height) {
+        return Error::UNSUPPORTED_RESOLUTION;
+    }
+
+    // Record the frame size (in bytes).
+    session->image_size = (size_t) fmt.fmt.pix.sizeimage;
+
+    //
+    // Buffer Initialization.
+    //
+
+    // Allocate and clear a requestbuffers struct, since we need it later.
+    struct v4l2_requestbuffers req;
+    memset(&req, 0, sizeof(req));
+
+    // Number of buffers to register.
+    req.count = NUM_BUFFERS;
+    req.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+    // Specify the memory mode. We are using "user pointer" mode.
+    // We will supply pointers to buffers we allocate.
+    req.memory = V4L2_MEMORY_USERPTR;
+
+    // Request the ability to use our own buffers.
+    if (ioctl(session->fd, VIDIOC_REQBUFS, &req) != 0) {
+        return Error::REQUEST_BUFFERS;
+    }
+
+    // Allocate our buffers.
+    for (size_t i = 0; i < NUM_BUFFERS; i++) {
+        session->buffers[i].size = session->image_size;
+        session->buffers[i].data = new uint8_t[session->image_size];
+    }
+
+    // Link each buffer with the driver.
+    for (size_t i = 0; i < NUM_BUFFERS; i++) {
```

Comment:
I didn't write this up before, but `size_t` concerns me a bit. Have you considered being more specific on what the size you need is?

Reply:
It should probably be `uint32_t` since that is the field type of `v4l2_buffer.index`, which is where it is assigned during the loop.